In [ ]:
# Import numpy library
import numpy as np

# Load pandas library- the common practice is to assign an alias pd
import pandas as pd

# Series and DataFrame Creation in Pandas

## Series: **one-dimensional** labeled array capable of holding any data type.

In [ ]:
# carete a Pandas series simply by passing a Python list to pd.Series()
s1 = pd.Series([11, 21, 31, 5, 6, 7, 7])

print(s1)

0    11
1    21
2    31
3     5
4     6
5     7
6     7
dtype: int64


By default, `Series` is assigned an integer index, but it can be changed using the index parameter.

In [ ]:
# Specify index
s3 = pd.Series([11, 21, 31],
               index = ['March','April', 'May'])

print(s3)

March    11
April    21
May      31
dtype: int64




## DataFrame: a **two-dimensional** labeled data structure with columns of potentially different types.

In [ ]:
# # One of the most commonly used methods to do so is with a Python dictionary.
# Create a dictionary
data = {'City': ['San Francisco', 'San Jose','Seattle'],
        'Population': [131000, 40000, 60000]}

data

{'City': ['San Francisco', 'San Jose', 'Seattle'],
 'Population': [131000, 40000, 60000]}

In [ ]:
# We simply pass this dictionary to the DataFrame constructor
df = pd.DataFrame(data)

print(df)

            City  Population
0  San Francisco      131000
1       San Jose       40000
2        Seattle       60000


* Dictionary keys become the column names and the values become the data stored in the DataFrame.
* We now have a DataFrame with 3 rows and 2 columns.

### We can also create a DataFrame from a numpy array

In [ ]:
# create a 5 x 3 numpy array, and the values are from a standard normal distribution
arr_2d = np.random.randn(5, 3)

arr_2d

array([[ 0.25662514,  1.29953433,  0.54344069],
       [ 1.04550375,  1.23227767, -0.16471082],
       [ 1.42279564,  1.03798489,  0.15638583],
       [-0.22544984,  0.10667776,  0.54182657],
       [ 1.42820124, -0.60074353, -0.48757085]])

In [ ]:
# We will get column labels by default
df = pd.DataFrame(arr_2d)

df

0         1         2
0  0.256625  1.299534  0.543441
1  1.045504  1.232278 -0.164711
2  1.422796  1.037985  0.156386
3 -0.225450  0.106678  0.541827
4  1.428201 -0.600744 -0.487571

In [ ]:
# We will add column names for this DataFrame
df = pd.DataFrame(arr_2d, columns = ['A', 'B', 'C'])

df

A         B         C
0  0.256625  1.299534  0.543441
1  1.045504  1.232278 -0.164711
2  1.422796  1.037985  0.156386
3 -0.225450  0.106678  0.541827
4  1.428201 -0.600744 -0.487571

In [ ]:
# Check the dataframe's shape
df.shape

(5, 3)

In [ ]:
# Check this dataframe's size
df.size

15

### Create a DataFrame by reading Data from a CSV File

### Loading a CSV File

In [ ]:
# Create URL - This is the train csv file from github
csv_url = 'https://raw.githubusercontent.com/agconti/kaggle-titanic/master/data/train.csv'

In [ ]:
# Load dataset - You can also read a csv file from your local drive
df = pd.read_csv(csv_url)

# deault "header = 0"

In [ ]:
# View first 5 rows
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Filtering a Data Frame
* Filtering with `loc` and `iloc` methods
* Filtering by Selecting a Subset of Columns
* Filtering by condition



In [ ]:
# Create a DataFrame
data = {'Name': ['Lily', 'Emma', 'John'],
        'Age': [38, 15, 28],
        'Education': ['high school', 'phd', 'bachelor'],
        'Gender': [0, 0, 1]}

customer = pd.DataFrame(data)

customer

Name  Age    Education  Gender
0  Lily   38  high school       0
1  Emma   15          phd       0
2  John   28     bachelor       1

## Filtering with `loc` and `iloc` methods

In [ ]:
# Let’s first use the `loc` method to select the first 2 rows and "Age", "Education" columns in the "customer" dataframe
# loc[row label, column label]

customer.loc[ : 1, ["Age","Education"]]

Age    Education
0   38  high school
1   15          phd

* The `:1` is the equivalent of `0:1` and it indicates the rows starting from 0 to 1.
* The column names are passed as a list to the `loc` method.



In [ ]:
# Let’s do the same operation using the `iloc` method.
# loc[row index, column index]
# start index(inclusive) : end index (exclusive)

customer.iloc[ : 2,  1 : 3]

Age    Education
0   38  high school
1   15          phd

## Selecting a Subset of Columns
* Select a single column
* Select multiple columns


In [ ]:
# Select column "Education"
customer['Education']

0    high school
1            phd
2       bachelor
Name: Education, dtype: object

In [ ]:
# Select column "Age", "Education"
customer[['Education', 'Age']]

Education  Age
0  high school   38
1          phd   15
2     bachelor   28

## What if we want to access a specific cell by row and column labels?
* `df.at[]`

In [ ]:
# a specific cell by row and column labels
customer.at[0, 'Education']

'high school'

## Filtering by condition(s)
* based on a conditon
* based on multiple conditions


In [ ]:
# Create a dictionary
data = {'Name': ['Lily', 'Emma', 'John'],
        'Age': [38, 15, 28],
        'Education': ['high school', 'phd', 'bachelor'],
        'SexCode': [0, 0, 1]}

# Now the row indexes are integers
df = pd.DataFrame(data)

df

Name  Age    Education  SexCode
0  Lily   38  high school        0
1  Emma   15          phd        0
2  John   28     bachelor        1

## One condition

In [ ]:
# based on a condition: Filter customers whose name is 'Lily'
filter = customer['Name'] == 'Lily'

customer_filtered1 = customer[filter]

customer_filtered1

Name  Age    Education  Gender
0  Lily   38  high school       0

In [ ]:
# The same as
customer_filtered2 = customer[customer['Name'] == 'Lily']

customer_filtered2

Name  Age    Education  Gender
0  Lily   38  high school       0

## Multiple conditions

In [ ]:
# Age >= 18, and female
filter = (customer['Age'] >= 18) & (customer['Gender'] == 0)


In [ ]:
customer_filtered = customer[filter]

customer_filtered

Name  Age    Education  Gender
0  Lily   38  high school       0

In [ ]:
customer

Name  Age    Education  Gender
0  Lily   38  high school       0
1  Emma   15          phd       0
2  John   28     bachelor       1

# Handling Missing Values with Pandas

## Missing Value Types and Representation

In [ ]:
# The Pandas library accepts both Python’s None and NumPy’s np.nan as missing values,
# so we can use both to indicate the missing values.

df = pd.DataFrame({
    "A": [1, 2, 3, np.nan],
    "B": [2.4, 6.2, 5.1, np.nan],
    "C": ["foo","zoo","bar", None]
})


print(df)

     A    B     C
0  1.0  2.4   foo
1  2.0  6.2   zoo
2  3.0  5.1   bar
3  NaN  NaN  None


# Finding the Missing Values
* `isna()` function
* `notna()` function

In [ ]:
df

A    B     C
0  1.0  2.4   foo
1  2.0  6.2   zoo
2  3.0  5.1   bar
3  NaN  NaN  None

In [ ]:
# Find out what elemnt is missing
df.isna()

A      B      C
0  False  False  False
1  False  False  False
2  False  False  False
3   True   True   True

## Challenge: Get # of missing values in each column

In [ ]:
# Get # of missing values in each column
df.isna().sum()


# Get # of missing values in each row
# df.isna().sum(axis=1)

A    1
B    1
C    1
dtype: int64

## Challenge: Get total # of missing values in this DataFrame

In [ ]:
# Get # of missing values in this DataFrame
df.isna().sum().sum()

3

In [ ]:
df.isna()

A      B      C
0  False  False  False
1  False  False  False
2  False  False  False
3   True   True   True

In [ ]:
# Find out what elemnt is missing
df.notna()

A      B      C
0   True   True   True
1   True   True   True
2   True   True   True
3  False  False  False

## Challenge: Count the number of non-missing values in the columns

In [ ]:
df.notna().sum()

A    3
B    3
C    3
dtype: int64

## Challenge: Get total # of non-missing values in this DataFrame

In [ ]:
df.notna().sum().sum()

9

# Dropping Rows and Columns with Missing Values
* `dropna()`

In [ ]:
df = pd.DataFrame({
    "A": [1, 2, 3, 4, 7],
    "B": [2.4, np.nan, 5.1, np.nan, 2.6],
    "C": ["phd", "phd","high school","high school", np.nan],
    "D": [3, np.nan, None, None, None]
})

df

A    B            C    D
0  1  2.4          phd  3.0
1  2  NaN          phd  NaN
2  3  5.1  high school  NaN
3  4  NaN  high school  NaN
4  7  2.6          NaN  NaN

In [ ]:
# Drop rows that have at least one missing value
df.dropna(axis = 0, how = "any")

A    B    C    D
0  1  2.4  phd  3.0

In [ ]:
# Keep rows that have at least 3 non-missing values
df.dropna(axis = 0, thresh = 3)

A    B            C    D
0  1  2.4          phd  3.0
2  3  5.1  high school  NaN

In [ ]:
# Another example
df = pd.DataFrame({
    "A": [1, 2, 3, 4, 7],
    "B": [2.4, np.nan, 5.1, np.nan, 2.6],
    "C": ["phd", "phd","high school","high school", np.nan]
})

df

A    B            C
0  1  2.4          phd
1  2  NaN          phd
2  3  5.1  high school
3  4  NaN  high school
4  7  2.6          NaN

In [ ]:
# Drop columns that have at least one missing value
df.dropna(axis = 1, how = "any")

A
0  1
1  2
2  3
3  4
4  7

In [ ]:
# Keep columns that have at least 4 non-missing values
df.dropna(axis = 1, thresh = 4)

A            C
0  1          phd
1  2          phd
2  3  high school
3  4  high school
4  7          NaN

## Replacing the Missing Values
* the average value of the column
* the most frequent value in the column
* a random number


In [ ]:
df = pd.DataFrame({
    "account_length": [30, 20, 20, np.nan, 30],
    "gender": ['F', 'M', 'F', np.nan, 'F'],
    "education": ['high school', "high school","high school","high school", np.nan],
})

df

account_length gender    education
0            30.0      F  high school
1            20.0      M  high school
2            20.0      F  high school
3             NaN    NaN  high school
4            30.0      F          NaN

In [ ]:
# Replacing the Missing Values with the average value of the column "account_length"
value_to_replace = df["account_length"].mean()

value_to_replace
# 30 + 20 + 20 + 30 = 100
# 100 / 4 = 25

25.0

In [ ]:
# fill the missing value with the mean of column
df["account_length"].fillna(value_to_replace)

0    30.0
1    20.0
2    20.0
3    25.0
4    30.0
Name: account_length, dtype: float64

In [ ]:
# fill the missing value with the mean of column and change it inplace
df["account_length"].fillna(value_to_replace, inplace = True)

df

account_length gender    education
0            30.0      F  high school
1            20.0      M  high school
2            20.0      F  high school
3            25.0    NaN  high school
4            30.0      F          NaN

In [ ]:
# Replacing the Missing Values with the most frequent value in the column "gender"
value_to_replace2 = df["gender"].mode()[0]

# Note: if we don't add `[0]`, it will return a series not a single value

value_to_replace2

'F'

In [ ]:
# fill the missing value with the mode of column and change it inplace
df["gender"].fillna(value_to_replace2, inplace = True)

df

account_length gender    education
0            30.0      F  high school
1            20.0      M  high school
2            20.0      F  high school
3            25.0      F  high school
4            30.0      F          NaN

In [ ]:
# Let's say we know that last customer's education is "phd"
df["education"].fillna("phd", inplace = True)

df

account_length gender    education
0            30.0      F  high school
1            20.0      M  high school
2            20.0      F  high school
3            25.0      F  high school
4            30.0      F          phd

## Sorting in a Data Frame
* `sort_values()`: sort a DataFrame by one or more of its columns

In [ ]:
data = {
    'Date': ['2023-09-01', '2023-09-02', '2023-09-03', '2023-09-04'],
    'Product': ['Product A', 'Product B', 'Product A', 'Product C'],
    'Quantity Sold': [100, 150, 120, 80],
    'Revenue': [5000, 7500, 6000, 4000]
}

# Create a DataFrame
df = pd.DataFrame(data)

df

Date    Product  Quantity Sold  Revenue
0  2023-09-01  Product A            100     5000
1  2023-09-02  Product B            150     7500
2  2023-09-03  Product A            120     6000
3  2023-09-04  Product C             80     4000

In [ ]:
# We want to easily see proudct from highest revenue to the lowest
# Sort the DataFrame by 'Revenue' column in descending order
df.sort_values(by ='Revenue', ascending = False)


Date    Product  Quantity Sold  Revenue
1  2023-09-02  Product B            150     7500
2  2023-09-03  Product A            120     6000
0  2023-09-01  Product A            100     5000
3  2023-09-04  Product C             80     4000

In [ ]:
# Sample DataFrame
data = {
    'Student_Name': ['Alice', 'Bob', 'Eve', 'David', 'Carol'],
    'Age': [25, 30, 22, 24, 28],
    'Score': [95, 95, 92, 88, 88]
}

df = pd.DataFrame(data)

df

Student_Name  Age  Score
0        Alice   25     95
1          Bob   30     95
2          Eve   22     92
3        David   24     88
4        Carol   28     88

In [ ]:
# Sort Score from highest to lowest, we also want to sort the student age from smallest to largest
# Sort by 'Age' in ascending order and then by 'Score' in descending order
df.sort_values(by =['Score', 'Age'], ascending=[False, True])

Student_Name  Age  Score
0        Alice   25     95
1          Bob   30     95
2          Eve   22     92
3        David   24     88
4        Carol   28     88

## Data Manipulation with Pandas
* `drop()`
* `replace()`
* `apply()`

## `drop()`: Remove rows or columns from a DataFrame

In [ ]:
df = pd.DataFrame({
    "user_id": ["1232as", "1323wv", "134pdf", "342t6ff"],
    "account_length": [31, 0, 0, 23],
    "total_orders": [4, 2, 3, 4],
    "tenure": [np.nan, np.nan, np.nan, np.nan]
})

df

user_id  account_length  total_orders  tenure
0   1232as              31             4     NaN
1   1323wv               0             2     NaN
2   134pdf               0             3     NaN
3  342t6ff              23             4     NaN

In [ ]:
# drop column "user_id" since this column does not have much info
df.drop(['user_id'], axis = 1)

account_length  total_orders  tenure
0              31             4     NaN
1               0             2     NaN
2               0             3     NaN
3              23             4     NaN

In [ ]:
# We want to drop two columns 'user_id' and 'tenure' since tenure has all missing values

df.drop(['user_id', 'tenure'], axis = 1)

account_length  total_orders
0              31             4
1               0             2
2               0             3
3              23             4

In [ ]:
# Note that the original df does not change
# we assign a new variable to `df.drop(['user_id'], axis = 1)`
# OR specify `inplace` to True
df

user_id  account_length  total_orders  tenure
0   1232as              31             4     NaN
1   1323wv               0             2     NaN
2   134pdf               0             3     NaN
3  342t6ff              23             4     NaN

In [ ]:
# drop column "user_id" and made the change inplace

df.drop(['user_id'], axis = 1, inplace = True)

In [ ]:
## Now the df changed
df

account_length  total_orders  tenure
0              31             4     NaN
1               0             2     NaN
2               0             3     NaN
3              23             4     NaN

## `replace()`: Replace values within a DataFrame

In [ ]:
df

account_length  total_orders  tenure
0              31             4     NaN
1               0             2     NaN
2               0             3     NaN
3              23             4     NaN

* The 0 in column "account_length" looks not right:
 * Let's replace it to value 10

In [ ]:
# replace 0 in "account_length" to 10
df['account_length'].replace(0, 10, inplace = True)

df

account_length  total_orders  tenure
0              31             4     NaN
1              10             2     NaN
2              10             3     NaN
3              23             4     NaN

* The `inplace=True` argument ensures that the changes are applied directly to the DataFrame df without having to assign the result to a new variable.
* If you don't use `inplace=True`, you need to assign the result to a new variable.

## Challenge: Replace 0 in column "account_length" with the mean of non-zero values in "account_length"



In [ ]:
df = pd.DataFrame({
    "user_id": ["1232as", "1323wv", "134pdf", "342t6ff"],
    "account_length": [31, 0, 0, 23],
    "total_orders": [4, 2, 3, 4],
    "tenure": [np.nan, np.nan, np.nan, np.nan]
})

df

user_id  account_length  total_orders  tenure
0   1232as              31             4     NaN
1   1323wv               0             2     NaN
2   134pdf               0             3     NaN
3  342t6ff              23             4     NaN

In [ ]:
# replace 0 in "account_length" to 10
# Get all non-zeros in column "account_length"
non_zeros_in_account_length = df['account_length'][df['account_length'] != 0]


non_zeros_in_account_length

0    31
3    23
Name: account_length, dtype: int64

In [ ]:
# Get non-zeros mean
non_zeros_mean = non_zeros_in_account_length.mean()

non_zeros_mean

27.0

In [ ]:
df['account_length'].replace(0,  non_zeros_mean, inplace = True)

df

user_id  account_length  total_orders  tenure
0   1232as              31             4     NaN
1   1323wv              27             2     NaN
2   134pdf              27             3     NaN
3  342t6ff              23             4     NaN

## `apply()`: Applying a given function along an axis of a DataFrame or Series.

In [ ]:
df = pd.DataFrame({
    "user_id": ["1232as", "1323wv", "134pdf", "342t6ff"],
    "account_length": [31, 0, 0, 23],
    "total_orders": [4, 2, 3, 4],
    "tenure": [np.nan, np.nan, np.nan, np.nan]
})

df

user_id  account_length  total_orders  tenure
0   1232as              31             4     NaN
1   1323wv               0             2     NaN
2   134pdf               0             3     NaN
3  342t6ff              23             4     NaN

* We noticed a data issue: The values in "account_lengh" column were mistakenly substract 2. We want to fix this.

In [ ]:
# Modity the data in column "account_lengh" using apply() and lamda function

df['account_length'] = df['account_length'].apply(lambda x: x + 2)

In [ ]:
df

user_id  account_length  total_orders  tenure
0   1232as              33             4     NaN
1   1323wv               2             2     NaN
2   134pdf               2             3     NaN
3  342t6ff              25             4     NaN

## Sometimes we'll need to add a new column from current data

In [ ]:
df = pd.DataFrame({
    "user_id": ["1232as", "1323wv", "134pdf", "342t6ff"],
    "account_length": [31, 0, 0, 23],
    "total_orders": [4, 2, 3, 4],
    "total_amt": [156, 127, 135, 466]
})

df

user_id  account_length  total_orders  total_amt
0   1232as              31             4        156
1   1323wv               0             2        127
2   134pdf               0             3        135
3  342t6ff              23             4        466

## We want to add a column "avg_order_amt" to this DataFrame

In [ ]:
df["avg_order_amt"] = df["total_amt"] / df["total_orders"]

df

user_id  account_length  total_orders  total_amt  avg_order_amt
0   1232as              31             4        156           39.0
1   1323wv               0             2        127           63.5
2   134pdf               0             3        135           45.0
3  342t6ff              23             4        466          116.5

# Data Analysis with Pandas
* `groupby()`
* `agg()`
* `reset_index()`
* `rename()`
* `idxmax()`


## Business Question: What is the average rating by product category?

In [ ]:
df = pd.DataFrame({
    'product_category': ['baby', 'book', 'book', 'beauty', 'baby'],
    'product_price': [19.99, 12.49, 22, 49, 16],
    'rating': [5, 3, 3, 2, 4]
})

df

product_category  product_price  rating
0             baby          19.99       5
1             book          12.49       3
2             book          22.00       3
3           beauty          49.00       2
4             baby          16.00       4

In [ ]:
# avg rating of amazon product review data
df.groupby('product_category')['rating'].mean()

product_category
baby      4.5
beauty    2.0
book      3.0
Name: rating, dtype: float64

* The position of column names looks strange

In [ ]:
# Use reset_index() to convert the grouped index back to columns

df.groupby('product_category')['rating'].mean().reset_index()

product_category  rating
0             baby     4.5
1           beauty     2.0
2             book     3.0

* The `reset_index()` is used to remove the current index and restore grouped columns to a DataFrame, which can be particularly useful when you want to work with data in a more organized form.

### Now we notice that the third column still has the name 'rating', which is confusing.












In [ ]:
avg_rating_summary = df.groupby('product_category')['rating'].mean().reset_index()

avg_rating_summary

product_category  rating
0             baby     4.5
1           beauty     2.0
2             book     3.0

In [ ]:
# Use rename function, we want to do it in the same dataframe
avg_rating_summary.rename(columns= {'rating': 'avg_rating'}, inplace = True)

avg_rating_summary

product_category  avg_rating
0             baby         4.5
1           beauty         2.0
2             book         3.0

## Business Question: What is the average rating of the "baby" cateogry?


In [ ]:
# Select product_category == 'baby'
baby_category = df[df['product_category'] == 'baby']

baby_category

product_category  product_price  rating
0             baby          19.99       5
4             baby          16.00       4

In [ ]:
baby_category['rating'].mean()

4.5

## Business Question: What is total amount of each order?


In [ ]:
df = pd.DataFrame({
    'order_id': ['4235', '2342', '1234', '5325', '1342'],
    'product_category': ['baby', 'beauty', 'baby', 'beauty', 'beatuty'],
    'product': ['diaper', 'eye liner', 'diaper', 'eye shadow', 'face mask'],
    'product_price': [19.99, 12.49, 22, 49, 16],
    'quantity': [5, 3, 3, 2, 4]
})

df

order_id product_category     product  product_price  quantity
0     4235             baby      diaper          19.99         5
1     2342           beauty   eye liner          12.49         3
2     1234             baby      diaper          22.00         3
3     5325           beauty  eye shadow          49.00         2
4     1342          beatuty   face mask          16.00         4

In [ ]:
df['amt'] = df['product_price'] * df['quantity']

df

order_id product_category     product  product_price  quantity    amt
0     4235             baby      diaper          19.99         5  99.95
1     2342           beauty   eye liner          12.49         3  37.47
2     1234             baby      diaper          22.00         3  66.00
3     5325           beauty  eye shadow          49.00         2  98.00
4     1342          beatuty   face mask          16.00         4  64.00

##  Business Question: What is the avg proudct price and total sales amount of each product category?

In [ ]:
# groupby, agg
# in agg() - column you want to apply function to : function

df.groupby('product_category').agg({'product_price': 'mean', 'amt': 'sum'})

product_price     amt
product_category                       
baby                     20.995  165.95
beatuty                  16.000   64.00
beauty                   30.745  135.47

In [ ]:
# Again, let's add reset_index()
summary = df.groupby('product_category').agg({'product_price': 'mean', 'amt': 'sum'}).reset_index()

summary

product_category  product_price     amt
0             baby         20.995  165.95
1          beatuty         16.000   64.00
2           beauty         30.745  135.47

In [ ]:
# We need to remame these two columns to make columns clear
# Use rename function, we want to do it in the same dataframe
summary.rename(columns= {'product_price': 'avg_product_price', 'amt' : 'revenue'}, inplace = True)

summary

product_category  avg_product_price  revenue
0             baby             20.995   165.95
1          beatuty             16.000    64.00
2           beauty             30.745   135.47

## Business Question: What's the product category that has the highest product price?
* Not use `sort_values()` function


In [ ]:
df = pd.DataFrame({
    'order_id': ['4235', '2342', '1234', '5325', '1342'],
    'product_category': ['baby', 'beauty', 'baby', 'beauty', 'beatuty'],
    'product': ['diaper', 'eye liner', 'diaper', 'eye shadow', 'face mask'],
    'product_price': [19.99, 12.49, 22, 49, 16],
    'quantity': [5, 3, 3, 2, 4]
})

df

order_id product_category     product  product_price  quantity
0     4235             baby      diaper          19.99         5
1     2342           beauty   eye liner          12.49         3
2     1234             baby      diaper          22.00         3
3     5325           beauty  eye shadow          49.00         2
4     1342          beatuty   face mask          16.00         4

## `idxmax()`: Return index of first occurrence of maximum over requested axis.
* Default value of `axis` is 0 : row

In [ ]:
# Use idxmax() function: axis=0
max_idx = df['product_price'].idxmax()

max_idx

3

In [ ]:
# Use loc function to select this observation
df.loc[max_idx]

order_id                  5325
product_category        beauty
product             eye shadow
product_price             49.0
quantity                     2
Name: 3, dtype: object

In [ ]:
# 'Beauty' category has the highest product price
df.loc[max_idx]['product_category']

'beauty'

## Business Question: What is the average number of orders by gender?

In [ ]:
df = pd.DataFrame({
    'user_id': ['4235', '2342', '1234', '5325', '1342'],
    'gender': ['F', 'F', 'F', 'M', 'M'],
    'total_order': [15, 13, 13, 12, 14]
})

df

user_id gender  total_order
0    4235      F           15
1    2342      F           13
2    1234      F           13
3    5325      M           12
4    1342      M           14

In [ ]:
# By survived with no `reset_index()`
df.groupby('gender')['total_order'].mean().reset_index()

gender  total_order
0      F    13.666667
1      M    13.000000

In [ ]:
# To make the column name right
df.groupby('gender')['total_order'].mean().reset_index().rename({'total_order' : 'avg_num_of_order'})

gender  total_order
0      F    13.666667
1      M    13.000000

# Combining DataFrames with Pandas

##  1. Concatenating DataFrames



In [ ]:
# Create URL - This is the train csv file from github
csv_url = 'https://raw.githubusercontent.com/agconti/kaggle-titanic/master/data/train.csv'

# Load dataset - You can also read a csv file from your local drive
titanic = pd.read_csv(csv_url)

titanic.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
# Suppose the titanic data is too large and stored in two data sources.
titanic_a = titanic.head(5)

titanic_a

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
titanic_b = titanic.tail(3)

titanic_b

PassengerId  Survived  Pclass                                      Name  \
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

## Now we want to concatenate these two small data frames to a titanic data sample
* Concatenate them by rows or column?

In [ ]:
# stack, axis = 0 -> concate along the row
concat_df = pd.concat([titanic_a, titanic_b], axis=0)

concat_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q

## Note the row index jump from 4 to 888, how shall we fix it?

In [ ]:
# Again, we use `reset_index()` function
concat_df.reset_index()

index  PassengerId  Survived  Pclass  \
0      0            1         0       3   
1      1            2         1       1   
2      2            3         1       3   
3      3            4         1       1   
4      4            5         0       3   
5    888          889         0       3   
6    889          890         1       1   
7    890          891         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
6                              Behr, Mr. Karl Howell    male  26.0      0   
7                                Dooley, Mr. Patrick    male  32.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      2        W./C. 6607  23.4500   NaN        S  
6      0            111369  30.0000  C148        C  
7      0            370376   7.7500   NaN        Q

## Concatenate data frame by column

In [ ]:
titanic_c = titanic[['Name', 'Age']].head(5)

titanic_c

Name   Age
0                            Braund, Mr. Owen Harris  22.0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
2                             Heikkinen, Miss. Laina  26.0
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
4                           Allen, Mr. William Henry  35.0

In [ ]:
titanic_d = titanic[['Sex', 'Fare']].head(3)

titanic_d

Sex     Fare
0    male   7.2500
1  female  71.2833
2  female   7.9250

In [ ]:
# Concatenate two data frames by column
pd.concat([titanic_c, titanic_d], axis=1)

Name   Age     Sex     Fare
0                            Braund, Mr. Owen Harris  22.0    male   7.2500
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0  female  71.2833
2                             Heikkinen, Miss. Laina  26.0  female   7.9250
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0     NaN      NaN
4                           Allen, Mr. William Henry  35.0     NaN      NaN

# Merging DataFrames - We want to merge two data frames


In [ ]:
# Sample product data
product_data = {
    'product_id': ['P1', 'P2', 'P3', 'P4'],
    'product_name': ['Product A', 'Product B', 'Product C', 'Product D']
}
product_df = pd.DataFrame(product_data)

product_df

product_id product_name
0         P1    Product A
1         P2    Product B
2         P3    Product C
3         P4    Product D

In [ ]:
# Sample review data
review_data = {
    'product_id': ['P1', 'P2', 'P3'],
    'review_score': [4.5, 3.8, 4.0],
    'review_comment': ['Great product', 'Needs improvement', 'Satisfied']
}

review_df = pd.DataFrame(review_data)

review_df

product_id  review_score     review_comment
0         P1           4.5      Great product
1         P2           3.8  Needs improvement
2         P3           4.0          Satisfied

## Now we want to merge these two data frames based on a common column.
* Which column shall we use?

In [ ]:
# Merge based on the 'product_id' column
merged_df = product_df.merge(review_df, on ='product_id')

merged_df

product_id product_name  review_score     review_comment
0         P1    Product A           4.5      Great product
1         P2    Product B           3.8  Needs improvement
2         P3    Product C           4.0          Satisfied

* Why does the above `merge` function only return 3 rows?
* `merge` defaults to inner joins, which only returns matching records.
* If we want to do an left join or right join, we can specify that with the **how** parameter:

In [ ]:
# Merge based on the 'product_id' column, left join
left_join_df = product_df.merge(review_df, on ='product_id', how = 'left')

left_join_df

product_id product_name  review_score     review_comment
0         P1    Product A           4.5      Great product
1         P2    Product B           3.8  Needs improvement
2         P3    Product C           4.0          Satisfied
3         P4    Product D           NaN                NaN

In [ ]:
# Merge based on the 'product_id' column, right join
right_join_df = product_df.merge(review_df, on='product_id', how = 'right')

right_join_df

product_id product_name  review_score     review_comment
0         P1    Product A           4.5      Great product
1         P2    Product B           3.8  Needs improvement
2         P3    Product C           4.0          Satisfied